tempodb/encoding/vparquet5/block_traceql.go

```diff
@@ -0,0 +1,3672 @@
+		return nil, nil
+	}
+
+	return parquetquery.NewLeftJoinIterator(DefinitionLevelResourceSpansILSSpanEvent, required, linkIters, linkCol, parquetquery.WithPool(pqLinkPool))
+}
+
+// createSpanIterator iterates through all span-level columns, groups them into rows representing
+// one span each.  Spans are returned that match any of the given conditions.
+func createSpanIterator(makeIter makeIterFn, innerIterators []parquetquery.Iterator, conditions []traceql.Condition, allConditions bool, dedicatedColumns backend.DedicatedColumns, selectAll bool,
+	sampler traceql.Sampler,
+) (parquetquery.Iterator, error) {
+	var (
+		columnSelectAs          = map[string]string{}
+		columnPredicates        = map[string][]parquetquery.Predicate{}
+		iters                   []parquetquery.Iterator
+		genericConditions       []traceql.Condition
+		columnMapping           = dedicatedColumnsToColumnMapping(dedicatedColumns, backend.DedicatedColumnScopeSpan)
+		nestedSetLeftExplicit   = false
+		nestedSetRightExplicit  = false
+		nestedSetParentExplicit = false
+	)
+
+	// todo: improve these methods. if addPredicate gets a nil predicate shouldn't it just wipe out the existing predicates instead of appending?
+	// nil predicate matches everything. what's the point of also evaluating a "real" predicate?
+	addPredicate := func(columnPath string, p parquetquery.Predicate) {
+		columnPredicates[columnPath] = append(columnPredicates[columnPath], p)
+	}
+
+	addNilPredicateIfNotAlready := func(path string) {
+		preds := columnPredicates[path]
+		foundOpNone := false
+
+		// check to see if there is a nil predicate and only add if it doesn't exist
+		for _, pred := range preds {
+			if pred == nil {
+				foundOpNone = true
+				break
+			}
+		}
+
+		if !foundOpNone {
+			addPredicate(path, nil)
+			columnSelectAs[path] = path
+		}
+	}
+
+	for _, cond := range conditions {
+		// Intrinsic?
+		switch cond.Attribute.Intrinsic {
+		case traceql.IntrinsicSpanID:
+			pred, err := createBytesPredicate(cond.Op, cond.Operands, true)
+			if err != nil {
+				return nil, err
+			}
+			addPredicate(columnPathSpanID, pred)
+			columnSelectAs[columnPathSpanID] = columnPathSpanID
+			continue
+
+		case traceql.IntrinsicParentID:
+			pred, err := createBytesPredicate(cond.Op, cond.Operands, true)
+			if err != nil {
+				return nil, err
+			}
+			addPredicate(columnPathSpanParentSpanID, pred)
+			columnSelectAs[columnPathSpanParentSpanID] = columnPathSpanParentSpanID
+			continue
+
+		case traceql.IntrinsicSpanStartTime:
+			// TODO - We also need to scale the operands if using lower precision.
+			pred, err := createIntPredicate(cond.Op, cond.Operands)
+			if err != nil {
+				return nil, err
+			}
+
+			if sampler != nil {
+				pred = newSamplingPredicate(sampler, pred)
+				// Removed so that it's not used down below.
+				sampler = nil
+			}
+
+			// Choose the least precise column possible.
+			// The step interval must be an even multiple of the pre-rounded precision.
```

Comment:
I assume `cond.Precision` is the one calculated in `engine_metrics.go:L992`, right? I'm curious about the edge cases here: is it possible that `cond.Precision` is larger 1m but still falls through to the high precision column due to misalignment?

Reply:
The engine now only sets Precision when the request is perfectly aligned:  https://github.com/grafana/tempo/pull/5495/files#diff-1b84537244dd74c3ba117a9181d6b27c2ba7d632b26c355ad5e5b428b6d00d19R993    Previously it only looked at the Step, which did have those gaps.